internal/rest/resources/hooks.go

```diff
@@ -0,0 +1,89 @@
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/canonical/lxd/lxd/response"
+	"github.com/gorilla/mux"
+
+	"github.com/canonical/microcluster/internal/rest/access"
+	"github.com/canonical/microcluster/internal/rest/client"
+	"github.com/canonical/microcluster/internal/rest/types"
+	"github.com/canonical/microcluster/internal/state"
+	"github.com/canonical/microcluster/rest"
+)
+
+var hooksCmd = rest.Endpoint{
+	Path: "hooks/{clusterMember}/{hookType}",
```

Comment:
Is there a requirement for having the `clusterMember` in the URL? 
Can't we call `hooks/{hookType}` on every member of the cluster directly if we already know the name (`clusterMember`) on which the hook should be executed? 

Below you do `if s.Name() != name {}` so do we require this "hidden" forwarding of the request if we could just call out to the member directly?

Reply:
That's too restrictive. It boxes us into an "all or nothing" approach. For example, `PreRemove` only runs on the to-be-removed node, but can be triggered remotely by any node that received the request to remove that node.

Reply:
Ah I think I misunderstood you, did you mean we should drop the name and the forwarding and just implicitly run the hook on the addresses we care about?

I was mainly including the node name to make it clear that POSTing to this API is node-specific, but if this isn't helpful then yes we can remove it. 

Actually the forwarding should be removed in general because microcluster has built-in forwarding that I forgot about. So if we do keep the node name we would just use it for validation and error out if it doesn't match.

Reply:
Jup that was the idea after I saw there is a loop calling out to each of the members directly. So I was wondering if there even is a situation in which we would call this endpoint on a specific member but the request would be forwarded?

Reply:
The forwarding is definitely unnecessary because microcluster actually has by-name forwarding built into every endpoint with a flag on the EndpointHandler. 

So if we do keep the name, it would just be to validate that we are on the right system and otherwise error out.

Reply:
It's superfluous right now because the API is minimal, but we should be ready for future changes and make them easy to implement. For example, how would we reconcile a new cluster-wide hooks API with the existing one?

Right now, we need to do this:
```
// Run the hook on only these relevant cluster members. This endpoint only accepts PreRemove, PostRemove, OnNewMember.
for member in releventClusterMembers {
  POST /cluster/internal/hooks/{member}/{hookType}
}
```

But what if in the future we want to do this:
```
 // Run certain hooks on all cluster members. This endpoint only accepts cluster-wide hooks, hypothetically like OnHeartbeat, OnDqliteRoleChange.
 POST /cluster/internal/hooks/{hookType}
```

Reply:
Another reason we might want to reserve the top level `hooks` API is if we ever want to have state for the hooks, like turning them off or on, and preloading data to be used with them.

Reply:
If we just have the `POST /cluster/internal/hooks/{hookType}` endpoint that triggers `hookType` only on the member where the endpoint is called, would we need any more in the future?

Having this single endpoint we can run the loop you have put as an example for each of the relevant cluster members since we also know it's address:

```
for member in releventClusterMembers {
  POST {address of relevant member}/cluster/internal/hooks/{hookType}
}
```

Do I miss anything here? Sorry if I am not following along with what you are describing.

Reply:
Actually, thinking more about this I think you're right, it's a bit unnecessary to keep the name around. As part of the built-in forwarding mechanism I mentioned above, we apply a query parameter `?target={clusterMember}` so if we need to make a distinction between global and per-node actions, we can leverage the query parameter.